athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneMetadataHandler.java

```diff
@@ -243,4 +271,63 @@ protected Field convertField(String name, String glueType)
     }
+
+    @Override
+    public GetTableResponse doGetQueryPassthroughSchema(BlockAllocator allocator, GetTableRequest request) throws Exception
+    {
+        Map<String, String> qptArguments = request.getQueryPassthroughArguments();
+        queryPassthrough.verify(qptArguments);
+        String schemaName = qptArguments.get(NeptuneQueryPassthrough.DATABASE);
+        String tableName = qptArguments.get(NeptuneQueryPassthrough.COLLECTION);
+        TableName tableNameObj = new TableName(schemaName, tableName);
+        request = new GetTableRequest(request.getIdentity(), request.getQueryId(),
+                request.getCatalogName(), tableNameObj, request.getQueryPassthroughArguments());
+
+        GetTableResponse getTableResponse = doGetTable(allocator, request);
+        List<Field> fields = getTableResponse.getSchema().getFields();
+        Client client = neptuneConnection.getNeptuneClientConnection();
+        GraphTraversalSource graphTraversalSource = neptuneConnection.getTraversalSource(client);
+        String gremlinQuery = qptArguments.get(NeptuneQueryPassthrough.QUERY);
+        gremlinQuery = gremlinQuery.concat(".limit(1)");
+        logger.info("NeptuneMetadataHandler doGetQueryPassthroughSchema gremlinQuery with limit: " + gremlinQuery);
+        ScriptEngine engine = new GremlinGroovyScriptEngine();
+        Bindings bindings = engine.createBindings();
+        bindings.put("g", graphTraversalSource);
+        Object object = engine.eval(gremlinQuery, bindings);
+        GraphTraversal graphTraversalForSchema = (GraphTraversal) object;
+        Map graphTraversalObj = null;
+        Schema schema;
+        while (graphTraversalForSchema.hasNext()) {
+            graphTraversalObj = (Map) graphTraversalForSchema.next();
+        }
+
+        //In case of gremlin query is fetching all columns then we can use same schema from glue
+        //otherwise we will build schema from gremlin query result column names.
+        if (graphTraversalObj != null && graphTraversalObj.size() == fields.size()) {
+            schema = getTableResponse.getSchema();
+        }
+        else {
+            schemaBuilder = SchemaBuilder.newBuilder();
+            //Building schema from gremlin query results and list of fields from glue response.
+            //It's require only when we are selecting limited columns.
+            graphTraversalObj.forEach((columnName, columnValue) -> buildSchema(columnName.toString(), fields));
+            Map<String, String> metaData = getTableResponse.getSchema().getCustomMetadata();
+            for (Map.Entry<String, String> map : metaData.entrySet()) {
+                schemaBuilder.addMetadata(map.getKey(), map.getValue());
+            }
+            schema = schemaBuilder.build();
+        }
+
+        return new GetTableResponse(request.getCatalogName(), tableNameObj, schema);
+    }
+
+    private void buildSchema(String columnName, List<Field> fields)
```

Comment:
instead of having line 98 here, can we just pass it to this method and modify the Schema builder?

Reply:
Hi @chngpe Thank you, Done code changes as per your suggestion.